From: SeongJae Park <sj@kernel.org>
Subject: mm/damon/sysfs-schemes: Fix leaking a filter for wrong cgroup path
Date: Mon, 19 Dec 2022 17:18:06 +0000

Commit f36f860207efa ("mm/damon/sysfs-schemes: implement scheme filters")
on mm-unstable introduced 'damon_syfs_set_scheme_filters()', which does
not free newly allocated filter when it fails setting memcg id for the
filter.  As a result, the memory for the filter leaks.  Fix it by freeing
the filter before returning the error.

Coverity CID: 1530032

Link: https://lkml.kernel.org/r/20221219171807.55708-2-sj@kernel.org
Fixes: f36f860207ef ("mm/damon/sysfs-schemes: implement scheme filters") on mm-unstable
Signed-off-by: SeongJae Park <sj@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/damon/sysfs-schemes.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/mm/damon/sysfs-schemes.c~mm-damon-sysfs-schemes-implement-scheme-filters-fix
+++ a/mm/damon/sysfs-schemes.c
@@ -1464,8 +1464,10 @@ static int damon_sysfs_set_scheme_filter
 			err = damon_sysfs_memcg_path_to_id(
 					sysfs_filter->memcg_path,
 					&filter->memcg_id);
-			if (err)
+			if (err) {
+				damos_destroy_filter(filter);
 				return err;
+			}
 		}
 		damos_add_filter(scheme, filter);
 	}
_
